packages/aws-cdk-lib/core/lib/stage.ts

```diff
@@ -210,16 +216,61 @@ export class Stage extends Construct {
    * calls will return the same assembly.
    */
   public synth(options: StageSynthesisOptions = { }): cxapi.CloudAssembly {
-    if (!this.assembly || options.force) {
+
+    let newConstructPaths = this.listAllConstructPaths(this);
+
+    // If the assembly cache is uninitiazed, run synthesize and reset construct paths cache
+    if (this.constructPathsCache.size == 0 || !this.assembly || options.force) {
       this.assembly = synthesize(this, {
         skipValidation: options.skipValidation,
         validateOnSynthesis: options.validateOnSynthesis,
       });
+      newConstructPaths = this.listAllConstructPaths(this);
+      this.constructPathsCache = newConstructPaths;
     }
 
+    // If the construct paths set has changed
+    if (!this.constructPathSetsAreEqual(this.constructPathsCache, newConstructPaths)) {
+      if ((options.errorOnDuplicateSynth ?? true) && !options.force) {
+        throw new Error('Synthesis has been called multiple times and the construct tree was modified after the first synthesis. This is not allowed: if you intended to do this, then call synthesis with the --force flag set to true.');
+      } else {
+        // eslint-disable-next-line no-console
+        console.error('Synthesis has been called mulitple times and the construct tree was modified after the first synthesis. Only the results of the first synth() call are used, and modifications done after it are ignored. Avoid construct tree mutations after synth() has been called unless this is intentional.');
```

Comment:
Why are we not using a construct warning here?

Reply:
Rico's suggestion: https://github.com/aws/aws-cdk/pull/31865#discussion_r1819425777